File: src/main/java/de/pbauerochse/worklogviewer/youtrack/YouTrackServiceFactory.java
package de.pbauerochse.worklogviewer.youtrack;

import com.google.common.collect.ImmutableSet;
import de.pbauerochse.worklogviewer.settings.Settings;
import de.pbauerochse.worklogviewer.settings.SettingsUtil;
import de.pbauerochse.worklogviewer.youtrack.v20173.YouTrackServiceV20173;
import de.pbauerochse.worklogviewer.youtrack.v20174.YouTrackServiceV20174;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Set;

import static de.pbauerochse.worklogviewer.util.ExceptionUtil.getIllegalArgumentException;

/**
 * Factory to get the YouTrackService
 * configured in the settings properties
 */
public class YouTrackServiceFactory {

    private static final Logger LOGGER = LoggerFactory.getLogger(YouTrackServiceFactory.class);

    private static final Set<YouTrackService> AVAILABLE_SERVICE_IMPLEMENTATIONS = ImmutableSet.of(
            new YouTrackServiceV20173(),
            new YouTrackServiceV20174()
    );

    private static YouTrackService cachedInstance = null;

    public static YouTrackService getInstance() {

        Settings settings = SettingsUtil.getSettings();

        if (authenticationMethodChanged(settings)) {
            cachedInstance = getYouTrackService(settings.getYouTrackVersion());

            LOGGER.info("Created new YouTrackService instance of type {}", cachedInstance.getClass().getSimpleName());
        }

        return cachedInstance;
    }

    public static YouTrackService getYouTrackService(YouTrackVersion version) {
        return AVAILABLE_SERVICE_IMPLEMENTATIONS.stream()
                .filter(connector -> connector.getVersion() == version)
                .findFirst()
                .orElseThrow(() -> getIllegalArgumentException("exceptions.settings.version.invalid", version.name()));
    }

    private static boolean authenticationMethodChanged(Settings settings) {
        return cachedInstance == null || !cachedInstance.getVersion().equals(settings.getYouTrackVersion());
    }

}

File: application/src/main/java/de/pbauerochse/worklogviewer/youtrack/YouTrackServiceFactory.kt
package de.pbauerochse.worklogviewer.youtrack

import de.pbauerochse.worklogviewer.settings.Settings
import de.pbauerochse.worklogviewer.settings.SettingsUtil
import de.pbauerochse.worklogviewer.util.ExceptionUtil.getIllegalArgumentException
import de.pbauerochse.worklogviewer.youtrack.v20174.UrlBuilder
import de.pbauerochse.worklogviewer.youtrack.v20174.YouTrackServiceV20174
import org.slf4j.LoggerFactory

/**
 * Factory to get the YouTrackConnector
 * configured in the settings properties
 */
object YouTrackServiceFactory {

    private val LOGGER = LoggerFactory.getLogger(YouTrackServiceFactory::class.java)
    private val AVAILABLE_SERVICE_IMPLEMENTATIONS = setOf(
        YouTrackServiceV20174(UrlBuilder(
            { SettingsUtil.settings.youTrackConnectionSettings.url },
            { SettingsUtil.settings.youTrackConnectionSettings.version }
        ))
    )

    private var cachedInstance: YouTrackService? = null

    @JvmStatic
    fun getYouTrackService(): YouTrackService {
        val settings = SettingsUtil.settings

        if (cachedInstance == null || authenticationMethodChanged(settings)) {
            cachedInstance = getYouTrackService(settings.youTrackConnectionSettings.version)
            LOGGER.info("Created new YouTrackConnector instance of type {}", cachedInstance!!.javaClass.simpleName)
        }

        return cachedInstance!!
    }

    private fun getYouTrackService(version: YouTrackVersion): YouTrackService {
        return AVAILABLE_SERVICE_IMPLEMENTATIONS.stream()
            .filter { service -> service.supportedVersions.contains(version) }
            .findFirst()
            .orElseThrow({ getIllegalArgumentException("exceptions.settings.version.invalid", version.name) })
    }

    private fun authenticationMethodChanged(settings: Settings): Boolean {
        return !cachedInstance!!.supportedVersions.contains(settings.youTrackConnectionSettings.version)
    }

}
